### ISSUE TITLE
Implement Dynamic Pricing AI Engine

### ISSUE THREAD

**Issue by caxtonacollins** [chioma-housing-protocol-I/chioma#526]
# Issue #086: Implement Dynamic Pricing AI Engine

**Community Link:** https://t.me/chiomagroup

## Problem Statement
Landlords need AI-powered pricing recommendations based on market demand, seasonality, competitor pricing, and historical data to maximize revenue while maintaining occupancy.

## Impact
- Helps landlords optimize pricing
- Increases platform revenue
- Improves occupancy rates
- Provides competitive advantage

### AI Model Architecture

```python
# backend/ai-services/pricing/pricing_model.py

import xgboost as xgb
import numpy as np
from datetime import datetime, timedelta
import pandas as pd

class DynamicPricingModel:
    def __init__(self):
        self.model = xgb.XGBRegressor(
            n_estimators=1000,
            learning_rate=0.01,
            max_depth=7,
            subsample=0.8,
            colsample_bytree=0.8,
            random_state=42
        )
        self.scaler = StandardScaler()
        
    def extract_features(self, property_data, date):
        """Extract features for pricing prediction"""
        features = {
            # Property features
            'bedrooms': property_data['bedrooms'],
            'bathrooms': property_data['bathrooms'],
            'square_feet': property_data['square_feet'],
            'latitude': property_data['latitude'],
            'longitude': property_data['longitude'],
            'amenity_count': len(property_data['amenities']),
            
            # Temporal features
            'day_of_week': date.weekday(),
            'month': date.month,
            'quarter': (date.month - 1) // 3 + 1,
            'is_weekend': 1 if date.weekday() >= 5 else 0,
            'is_holiday': self.is_holiday(date),
            'days_until_booking': (date - datetime.now()).days,
            'is_peak_season': self.is_peak_season(date),
            
            # Market features
            'avg_competitor_price': self.get_competitor_avg(property_data),
            'competitor_count': self.get_competitor_count(property_data),
            'occupancy_rate': self.get_market_occupancy(property_data),
            'demand_score': self.calculate_demand(property_data, date),
            'price_index': self.get_price_index(property_data, date),
            
            # Historical features
            'avg_price_last_30d': self.get_historical_avg(property_data, 30),
            'avg_price_last_90d': self.get_historical_avg(property_data, 90),
            'booking_rate_last_30d': self.get_booking_rate(property_data, 30),
            'cancellation_rate': self.get_cancellation_rate(property_data),
            'review_score': property_data.get('review_score', 4.5),
            
            # External factors
            'weather_score': self.get_weather_score(property_data, date),
            'event_proximity': self.get_event_proximity(property_data, date),
            'economic_index': self.get_economic_index(date)
        }
        
        return np.array(list(features.values()))
    
    def predict_optimal_price(self, property_id, date_range):
        """Predict optimal price for date range"""
        property_data = self.get_property_data(property_id)
        predictions = []
        
        for date in date_range:
            features = self.extract_features(property_data, date)
            features_scaled = self.scaler.transform([features])
            
            base_price = self.model.predict(features_scaled)[0]
            
            # Apply business rules
            min_price = property_data.get('min_price', base_price * 0.7)
            max_price = property_data.get('max_price', base_price * 1.5)
            optimal_price = np.clip(base_price, min_price, max_price)
            
            # Calculate confidence
            confidence = self.calculate_confidence(features)
            
            # Get demand level
            demand_level = self.get_demand_level(features)
            
            predictions.append({
                'date': date.isoformat(),
                'price': round(optimal_price, 2),
                'confidence': round(confidence, 2),
                'demand_level': demand_level,
                'recommendation': self.get_recommendation(optimal_price, property_data)
            })
        
        return predictions
    
    def suggest_mode_switch(self, property_id, days_ahead=90):
        """Suggest optimal rental mode"""
        property_data = self.get_property_data(property_id)
        
        # Predict revenue for next N days
        date_range = [datetime.now() + timedelta(days=i) for i in range(days_ahead)]
        
        short_term_revenue = self.predict_short_term_revenue(property_id, date_range)
        long_term_revenue = self.predict_long_term_revenue(property_id, days_ahead)
        
        recommendation = {
            'current_mode': property_data['rental_mode'],
            'suggested_mode': 'short_term' if short_term_revenue > long_term_revenue else 'long_term',
            'short_term_revenue': round(short_term_revenue, 2),
            'long_term_revenue': round(long_term_revenue, 2),
            'revenue_difference': round(short_term_revenue - long_term_revenue, 2),
            'confidence': self.calculate_mode_confidence(property_data),
            'reasoning': self.generate_reasoning(property_data, short_term_revenue, long_term_revenue)
        }
        
        return recommendation
    
    def calculate_confidence(self, features):
        """Calculate prediction confidence (0-1)"""
        # Higher confidence for dates closer to now
        days_ahead = features[12]  # days_until_booking
        base_confidence = 0.95 if days_ahead <= 30 else 0.85 if days_ahead <= 60 else 0.75
        
        # Adjust based on data availability
        historical_data_score = min(features[20] / 100, 1.0)  # booking_rate
        
        return base_confidence * (0.7 + 0.3 * historical_data_score)
    
    def get_demand_level(self, features):
        """Categorize demand level"""
        demand_score = features[14]  # demand_score
        
        if demand_score >= 0.8:
            return 'very_high'
        elif demand_score >= 0.6:
            return 'high'
        elif demand_score >= 0.4:
            return 'medium'
        elif demand_score >= 0.2:
            return 'low'
        else:
            return 'very_low'
    
    def get_recommendation(self, price, property_data):
        """Get pricing recommendation"""
        current_price = property_data.get('current_price', price)
        change_percent = ((price - current_price) / current_price) * 100
        
        if change_percent > 10:
            return 'increase_significantly'
        elif change_percent > 5:
            return 'increase'
        elif change_percent < -10:
            return 'decrease_significantly'
        elif change_percent < -5:
            return 'decrease'
        else:
            return 'keep_current'
```

### Backend API Endpoints

```typescript
// backend/src/modules/ai/controllers/pricing-ai.controller.ts

@Controller('api/ai/pricing')
@UseGuards(AuthGuard)
export class PricingAiController {
  constructor(private readonly pricingAiService: PricingAiService) {}

  // Get price predictions for date range
  @Get('predict/:propertyId')
  async predictPrice(
    @Param('propertyId') propertyId: string,
    @Query('startDate') startDate: string,
    @Query('endDate') endDate: string,
    @Req() req: Request
  ) {
    return this.pricingAiService.predictOptimalPrice(
      propertyId,
      new Date(startDate),
      new Date(endDate),
      req.user.id
    );
  }

  // Get mode recommendation
  @Get('mode-recommendation/:propertyId')
  async getModeRecommendation(
    @Param('propertyId') propertyId: string,
    @Query('daysAhead') daysAhead = 90,
    @Req() req: Request
  ) {
    return this.pricingAiService.suggestModeSwitch(propertyId, daysAhead, req.user.id);
  }

  // Apply pricing suggestion
  @Post('apply-suggestion/:propertyId')
  async applyPricingSuggestion(
    @Param('propertyId') propertyId: string,
    @Body() dto: ApplyPricingDto,
    @Req() req: Request
  ) {
    return this.pricingAiService.applyPricingSuggestion(propertyId, dto, req.user.id);
  }

  // Get pricing analytics
  @Get('analytics/:propertyId')
  async getPricingAnalytics(
    @Param('propertyId') propertyId: string,
    @Query('days') days = 90,
    @Req() req: Request
  ) {
    return this.pricingAiService.getPricingAnalytics(propertyId, days, req.user.id);
  }
}
```

### Service Implementation

```typescript
// backend/src/modules/ai/services/pricing-ai.service.ts

@Injectable()
export class PricingAiService {
  constructor(
    @InjectRepository(Property)
    private propertyRepo: Repository<Property>,
    private httpService: HttpService,
    private cacheService: CacheService
  ) {}

  async predictOptimalPrice(
    propertyId: string,
    startDate: Date,
    endDate: Date,
    userId: string
  ) {
    // Verify ownership
    const property = await this.propertyRepo.findOne(propertyId);
    if (!property || property.ownerId !== userId) {
      throw new ForbiddenException('Not authorized');
    }

    // Check cache
    const cacheKey = `pricing:${propertyId}:${startDate}:${endDate}`;
    const cached = await this.cacheService.get(cacheKey);
    if (cached) return cached;

    // Call Python AI service
    const predictions = await this.callAiService('predict_price', {
      propertyId,
      startDate,
      endDate,
      propertyData: property
    });

    // Cache for 24 hours
    await this.cacheService.set(cacheKey, predictions, 86400);

    return predictions;
  }

  async suggestModeSwitch(propertyId: string, daysAhead: number, userId: string) {
    const property = await this.propertyRepo.findOne(propertyId);
    if (!property || property.ownerId !== userId) {
      throw new ForbiddenException('Not authorized');
    }

    const recommendation = await this.callAiService('suggest_mode', {
      propertyId,
      daysAhead,
      propertyData: property
    });

    return recommendation;
  }

  private async callAiService(method: string, data: any) {
    try {
      const response = await this.httpService.post(
        `${process.env.AI_SERVICE_URL}/pricing/${method}`,
        data
      ).toPromise();
      return response.data;
    } catch (error) {
      throw new InternalServerErrorException('AI service error');
    }
  }
}
```

## Acceptance Criteria
- [ ] XGBoost model trained on historical data
- [ ] Feature extraction working correctly
- [ ] Price predictions accurate (within 10% of actual)
- [ ] Mode recommendations working
- [ ] API endpoints implemented
- [ ] Caching implemented for performance
- [ ] Model performance tracked

## Testing Instructions

```bash
# Train model
python backend/ai-services/pricing/train_model.py

# Test predictions
curl -X GET "http://localhost:3000/api/ai/pricing/predict/PROP_ID?startDate=2026-04-01&endDate=2026-04-30"

# Get mode recommendation
curl -X GET "http://localhost:3000/api/ai/pricing/mode-recommendation/PROP_ID?daysAhead=90"
```

### Test Cases
1. Predict prices for date range
2. Verify predictions within acceptable range
3. Get mode recommendations
4. Apply pricing suggestions
5. Test caching
6. Verify authorization

## Effort Estimation
- **Complexity:** Very High
- **Time:** 12-15 hours
- **Dependencies:** XGBoost, Python, FastAPI, Redis

## Notes
- Requires historical booking data for training
- Consider model retraining frequency
- Monitor model accuracy over time
- Plan for A/B testing pricing recommendations

**Comment by Jemiiah** [comment]
@Jemiiah has applied to work on this issue as part of the **Stellar Wave Program**'s 3rd wave.


> I would like to work on this


ℹ️ **Repo Maintainers**: To accept this application, [review their application](https://www.drips.network/wave/maintainers/issues/397ee279-fa42-44c3-9332-d610eed96c43) or assign @Jemiiah to this issue.
